pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -32,70 +32,141 @@ type Alertmanager struct {
 	// silences keeps the track of which notifications we should not fire due to user configuration.
 	silences   *silence.Silences
-	marker     types.Marker
 	alerts     *AlertProvider
 	dispatcher *dispatch.Dispatcher
 
+	configReloadC chan struct{}
+
 	wg sync.WaitGroup
 }
 
 func init() {
-	registry.RegisterService(&Alertmanager{})
+	registry.RegisterService(&Alertmanager{
+		configReloadC: make(chan struct{}, 1),
+	})
 }
 
 func (am *Alertmanager) IsDisabled() bool {
 	return !setting.AlertingEnabled || !setting.ExecuteAlerts
 }
 
-func (am *Alertmanager) Init() error {
+func (am *Alertmanager) Init() (err error) {
 	am.logger = log.New("alertmanager")
+	am.notificationLog, err = nflog.New(
+		nflog.WithRetention(time.Hour*24),                         //TODO: This is a setting.
+		nflog.WithSnapshot(filepath.Join("dir", "notifications")), //TODO: This should be a setting
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to initialize the notification log component of alerting")
+	}
 
 	return nil
 }
 
 func (am *Alertmanager) Run(ctx context.Context) error {
+	reloadFromConfig := func() error {
+		cfg1, cfg2, err := getConfigFromDatabase()
+		if err != nil {
+			return errors.Wrap(err, "get config from database")
+		}
+		return errors.Wrap(am.reloadFromConfig(cfg1, cfg2), "reload config")
+	}
+	// Make sure dispatcher starts. We can tolerate future reload failures.
+	if err := reloadFromConfig(); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				if am.dispatcher != nil {
+					am.dispatcher.Stop()
+				}
+				am.wg.Wait()
+			case <-time.After(1 * time.Minute):
+				am.TriggerConfigReload()
+			case <-am.configReloadC:
+				// TODO: add metrics for failure?
+				if err := am.syncWithDatabase(); err != nil {
+					am.logger.Error("failed to sync config from database", "error", err)
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (am *Alertmanager) TriggerConfigReload() {
+	select {
+	case am.configReloadC <- struct{}{}:
+	default:
+	}
+}
+
+func (am *Alertmanager) syncWithDatabase() error {
+	// TODO: check if config is same as before using hashes and skip reload in case they are same.
+	cfg1, cfg2, err := getConfigFromDatabase()
+	if err != nil {
+		return errors.Wrap(err, "get config from database")
+	}
+	return errors.Wrap(am.reloadFromConfig(cfg1, cfg2), "reload from config")
+}
+
+func getConfigFromDatabase() (interface{}, interface{}, error) {
+	// TODO: get configs from the database.
+	return nil, nil, nil
+}
+
+// reloadFromConfig re-initialises all components using the config provided and restarts dispatcher.
+// TODO: replace these dummy configs with actual ones and use them.
+func (am *Alertmanager) reloadFromConfig(cfg1 interface{}, cfg2 interface{}) error {
 	//TODO: Speak with David Parrot wrt to the marker, we'll probably need our own.
-	am.marker = types.NewMarker(prometheus.DefaultRegisterer)
+	marker := types.NewMarker(prometheus.DefaultRegisterer)
```

Comment:
The marker ad well, this is can just the one.

Reply:
Okay